### PR TITLE
ruby27: Add beginless ranges support

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

Add support for begin-less ranges as described [here](https://rubyreferences.github.io/rubychanges/2.7.html#beginless-range):

```ruby
g = (..1)
T.reveal_type(g) # error: Revealed type: `T::Range[Integer(1)]`

h = Range.new(nil, 1)
T.reveal_type(h) # error: Revealed type: `T::Range[Integer(1)]`
```

This commit tracks upstream commit ruby/ruby@95f7992.

### Test plan

See included automated tests.

### CODE REVIEW

File: test/testdata/desugar/range.rb [hunk 1]
```diff
@@ -14,9 +14,15 @@ def foo
...
+    g = (..1)
+    h = Range.new(nil, 1)
     T.reveal_type(g) # error: Revealed type: `T::Range[Integer(1)]`
     T.reveal_type(h) # error: Revealed type: `T::Range[Integer(1)]`
+
+    # testing for endless ranges
+    i = (1..)
+    j = Range.new(1, nil)
+    T.reveal_type(i) # error: Revealed type: `T::Range[Integer(1)]`
+    T.reveal_type(j) # error: Revealed type: `T::Range[Integer(1)]`
```
Comment (paracycle) [inline test/testdata/desugar/range.rb:27]: Can we also test for both beginless and endless ranges? They should give `T::Range[T.untyped]` since we obviously can't detect the type of the elements, but I think that was a problematic edge-case up to now. The two ways to create it are:
```ruby
(nil..)
Range.new(nil, nil)
```

Comment (paracycle) [reply test/testdata/desugar/range.rb:27]: Based on further conversation offline with @Morriar, we both think that it is more appropriate to model a limitless range (`Range.new(nil, nil)`) as `T::Range[T.noreturn]` since one cannot enumerate or even access random elements of the range. Thus, it should be a static error if anyone tried to access `.first` or `.last` on such a range.

Comment (Morriar) [reply test/testdata/desugar/range.rb:27]: Done :)
